### PR TITLE
Unit tests for BMP regression tool

### PR DESCRIPTION
**Pull request by ElliotFreebairn**
The unit tests require that **Catch2** [https://github.com/catchorg/Catch2](url) is available and built on your system

To run the unit tests:
`cd test && make check`

### CODE REVIEW

**Comment by ElliotFreebairn** [comment]
I've pretty much coverered most of the BMP code, so I just want to check that I'm on the right track before continuing on with the pixelbasher code

File: test/makefile [hunk 1]
```diff
@@ -0,0 +1,32 @@
+SRC_DIR = .
+OBJ_DIR = obj
+TARGET = unittest
+
+CXX = g++
+CXXFLAGS = -std=c++20 -Wall -g -MMD -MP -I/usr/local/include -I../src
```
Comment (stbergmann) [inline test/makefile]: What's the `-I/usr/local/include` needed for?  Finding Catch2 include files should be covered through the `Catch2_CFLAGS` below.

File: test/makefile [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+TEST_SRCS = $(wildcard $(SRC_DIR)/*.cpp)
+
+SRC_FILES = ../src/bmp.cpp ../src/pixelbasher.cpp
+
+SRCS = $(TEST_SRCS) $(SRC_FILES)
+
+OBJS = $(patsubst %.cpp, $(OBJ_DIR)/%.o, $(notdir $(SRCS)))
+
+$(OBJ_DIR)/%.o:
```
Comment (stbergmann) [inline test/makefile]: This is somewhat "dangerous":  Originally, the below `-include $(OBJS:.o=.d)` will not find the `*.d` files it is supposed to include, so it will search for rules that would create them.  There apparently is some build-in implicit rule (cf. `make -r` to suppress those) that tells how to make a `*.d` file from a `*.d.o` file, and to make a `*.d.o` file, it will use this rule (which doesn't have any prerequisites, so will work for any `*.o` file named whatsoever, including any file name `*.d.o`). But the below `$(filter %/$*.cpp, $(SRCS))` will produce nothing in that case, and `gcc` will fail with
```
g++: fatal error: no input files
compilation terminated.
```
Better use
```
vpath %.cpp ../src
```
to tell it to search for `*.cpp` files also in that dir (in addition to searching in this dir), and use a more specific rule
```
$(OBJ_DIR)/%.o: %.cpp
	mkdir -p $(OBJ_DIR)
	$(CXX) $(CXXFLAGS) $(Catch2_CFLAGS) -c $< -o $@
```
instead.

File: test/test_bmp.cpp [hunk 3]
```diff
@@ -0,0 +1,316 @@
+#include <catch2/catch.hpp>
```
Comment (stbergmann) [inline test/test_bmp.cpp]: I'm not sure what kind of version of Catch2 you are using?  I built one locally from recent `devel` branch of <https://github.com/catchorg/Catch2>, and that doesn't have any `catch2/catch.hpp` include file.  (It has some `catch2/catch_all.hpp`, but when I try with that instead, I get lots of
```
./test_bmp.cpp:49:68: error: ‘Contains’ is not a member of ‘Catch’; did you mean ‘contains’?
   49 |         REQUIRE_THROWS_WITH(not_bmp.read(bmp_path.c_str()), Catch::Contains("Not a BMP") && Catch::Contains("BM"));
      |                                                                    ^~~~~~~~
```
etc. errors.)

Comment (ElliotFreebairn) [reply test/test_bmp.cpp]: I installed catch2 from my package manager, which was my fault. I'm now building it from the recent `devel` branch

Comment (stbergmann) [reply test/test_bmp.cpp]: > I installed catch2 from my package manager, which was my fault. I'm now building it from the recent `devel` branch

oh, no idea whether that was a fault or a feature---I'm not sure either the tip of their (default) `devel` branch is the best thing to use, just wondered what the story there is, why my installation was apparently so fundamentally different than what you were using.  (i.e., I have no idea what version would be the best to use)

**Comment by ElliotFreebairn** [comment]
@stbergmann I've made the changes you've requested and migrated to the v3 Catch2. Hopefully it should be working for you, let me know :)

File: test/makefile [hunk 4]
```diff
@@ -3,30 +3,35 @@ OBJ_DIR = obj
...
--include $(OBJS:.o=.d)
+-include $(wildcard $(OBJS:.o=.d))
 .PHONY: clean check
 
 check: $(TARGET)
 	./$(TARGET)
 
 clean:
-	rm -fr $(OBJ_DIR) $(TARGET)
+	rm -fr $(OBJ_DIR) $(TARGET)
```
Comment (stbergmann) [inline test/makefile]: you dropped the final newline; I would put it back

File: test/makefile [hunk 5]
```diff
@@ -3,30 +3,35 @@ OBJ_DIR = obj
...
+$(OBJ_DIR)/%.o: ../src/%.cpp
 	mkdir -p $(OBJ_DIR)
-	$(CXX) $(CXXFLAGS) $(Catch2_CFLAGS) -c $(filter %/$*.cpp, $(SRCS)) -o $@
+	$(CXX) $(CXXFLAGS) $(Catch2_CFLAGS) -c $< -o $@
 
 $(TARGET): $(OBJS)
 	$(CXX) $(CXXFLAGS) $(Catch2_CFLAGS) $(OBJS) -o $(TARGET) $(Catch2_LDFLAGS)
 
--include $(OBJS:.o=.d)
+-include $(wildcard $(OBJS:.o=.d))
```
Comment (stbergmann) [inline test/makefile]: Is this use of `wildcard` really needed?  It looks somewhat hackish/too-clever, as it doesn't actually involve any patterns with wildcards...

**Review by stbergmann — COMMENTED** [review]
Without having looked at your actual unit test code much (yet):
* Nothing creates a `test/test_data/output` dir, so all tests wanting to write there fail for me.
* There's no `test/test_data/not_bmp.png` and `test/test_data/24_bit.bmp`, so the tests reading those fail for me.
* What also fails for me is
```
-------------------------------------------------------------------------------
Writing filters & masks onto BMP files
  Writes and re-reads a BMP with all red due to edge mask
-------------------------------------------------------------------------------
test_bmp.cpp:90
...............................................................................

test_bmp.cpp:100: FAILED:
  REQUIRE( dummy_image_red_count == 0 )
with expansion:
  32764 (0x7ffc) == 0
```

**Comment by ElliotFreebairn** [comment]
> -------------------------------------------------------------------------------
> Writing filters & masks onto BMP files
>   Writes and re-reads a BMP with all red due to edge mask
> -------------------------------------------------------------------------------
> test_bmp.cpp:90
> ...............................................................................
> 
> test_bmp.cpp:100: FAILED:
>   REQUIRE( dummy_image_red_count == 0 )
> with expansion:
>   32764 (0x7ffc) == 0
>  ...............................................................................

Thank you for the comments

I cloned the repo and tested make check which Is working for me with new changes. Is it still failing for you?

**Comment by stbergmann** [comment]
> I cloned the repo and tested make check which Is working for me with new changes. Is it still failing for you?

Yes, with `HEAD` at latest `268912bbf41e1c7c89d0973f44d910b2f68c03f6` "Renaming read to input for better readaiblity", a from-scratch build of
```
$ PKG_CONFIG_PATH=/home/sberg/Catch2/inst/share/pkgconfig make -C test check
make: Entering directory '/home/sberg/github.com/ElliotFreebairn/image-regression-mso-test/test'
mkdir -p obj
g++ -std=c++20 -Wall -g -MMD -MP -I../src -I/home/sberg/Catch2/inst/include -c test_bmp.cpp -o obj/test_bmp.o
mkdir -p obj
g++ -std=c++20 -Wall -g -MMD -MP -I../src -I/home/sberg/Catch2/inst/include -c ../src/bmp.cpp -o obj/bmp.o
mkdir -p obj
g++ -std=c++20 -Wall -g -MMD -MP -I../src -I/home/sberg/Catch2/inst/include -c ../src/pixelbasher.cpp -o obj/pixelbasher.o
g++ -std=c++20 -Wall -g -MMD -MP -I../src -I/home/sberg/Catch2/inst/include  obj/test_bmp.o  obj/bmp.o  obj/pixelbasher.o -o unittest -L/home/sberg/Catch2/inst/lib64 -lCatch2Main -lCatch2
mkdir -p test_data/output/
./unittest
Randomness seeded to: 4085894767

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
unittest is a Catch2 v3.9.0 host application.
Run with -? for options

-------------------------------------------------------------------------------
Writing filters & masks onto BMP files
  Writes and re-reads a BMP with all red due to edge mask
-------------------------------------------------------------------------------
test_bmp.cpp:90
...............................................................................

test_bmp.cpp:100: FAILED:
  REQUIRE( dummy_image_red_count == 0 )
with expansion:
  32765 (0x7ffd) == 0

===============================================================================
test cases: 11 | 10 passed | 1 failed
assertions: 36 | 35 passed | 1 failed

make: *** [makefile:35: check] Error 42
make: Leaving directory '/home/sberg/github.com/ElliotFreebairn/image-regression-mso-test/test'
```
still fails for me.

**Comment by ElliotFreebairn** [comment]
> > I cloned the repo and tested make check which Is working for me with new changes. Is it still failing for you?
> 
> Yes, with `HEAD` at latest `268912bbf41e1c7c89d0973f44d910b2f68c03f6` "Renaming read to input for better readaiblity", a from-scratch build of
> 
> ```
> $ PKG_CONFIG_PATH=/home/sberg/Catch2/inst/share/pkgconfig make -C test check
> make: Entering directory '/home/sberg/github.com/ElliotFreebairn/image-regression-mso-test/test'
> mkdir -p obj
> g++ -std=c++20 -Wall -g -MMD -MP -I../src -I/home/sberg/Catch2/inst/include -c test_bmp.cpp -o obj/test_bmp.o
> mkdir -p obj
> g++ -std=c++20 -Wall -g -MMD -MP -I../src -I/home/sberg/Catch2/inst/include -c ../src/bmp.cpp -o obj/bmp.o
> mkdir -p obj
> g++ -std=c++20 -Wall -g -MMD -MP -I../src -I/home/sberg/Catch2/inst/include -c ../src/pixelbasher.cpp -o obj/pixelbasher.o
> g++ -std=c++20 -Wall -g -MMD -MP -I../src -I/home/sberg/Catch2/inst/include  obj/test_bmp.o  obj/bmp.o  obj/pixelbasher.o -o unittest -L/home/sberg/Catch2/inst/lib64 -lCatch2Main -lCatch2
> mkdir -p test_data/output/
> ./unittest
> Randomness seeded to: 4085894767
> 
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> unittest is a Catch2 v3.9.0 host application.
> Run with -? for options
> 
> -------------------------------------------------------------------------------
> Writing filters & masks onto BMP files
>   Writes and re-reads a BMP with all red due to edge mask
> -------------------------------------------------------------------------------
> test_bmp.cpp:90
> ...............................................................................
> 
> test_bmp.cpp:100: FAILED:
>   REQUIRE( dummy_image_red_count == 0 )
> with expansion:
>   32765 (0x7ffd) == 0
> 
> ===============================================================================
> test cases: 11 | 10 passed | 1 failed
> assertions: 36 | 35 passed | 1 failed
> 
> make: *** [makefile:35: check] Error 42
> make: Leaving directory '/home/sberg/github.com/ElliotFreebairn/image-regression-mso-test/test'
> ```
> 
> still fails for me.

hmm ok, sorry for the bother, I'll take a deeper look

**Comment by ElliotFreebairn** [comment]
@stbergmann I've had a crack at trying to fix that bug. It seemed to be that the BMP's constructor which creates an image based off height and width wasn't initializing header values properly, so that red count value might be garbage.

**Comment by stbergmann** [comment]
> @stbergmann I've had a crack at trying to fix that bug. It seemed to be that the BMP's constructor which creates an image based off height and width wasn't initializing header values properly, so that red count value might be garbage.

that was still failing for me, but I found the remaining issue now---it's the implementation of `BMP::calculate_colour_count` that calculates with uninitialized data :)

**Comment by ElliotFreebairn** [comment]
> > @stbergmann I've had a crack at trying to fix that bug. It seemed to be that the BMP's constructor which creates an image based off height and width wasn't initializing header values properly, so that red count value might be garbage.
> 
> that was still failing for me, but I found the remaining issue now---it's the implementation of `BMP::calculate_colour_count` that calculates with uninitialized data :)

Oh thank you so much, I don't think I would have looked at that. I'll push changes for that soon

**Comment by ElliotFreebairn** [comment]
> --it's the implementation of `BMP::calculate_colour_count` that calculates with uninitialized data :)

When you mean uninitialized data, is that for `int colour_count` not being `int colour_count = 0` or were you talking about `m_data`

I also realized that the indexing was incorrect while looking for a fix which will help.

**Comment by stbergmann** [comment]
> When you mean uninitialized data, is that for `int colour_count` not being `int colour_count = 0` or were you talking about `m_data`

the former

> I also realized that the indexing was incorrect while looking for a fix which will help.

(I didn't look at the actual code of this PR at all yet, beyond seeing it do a successful `make check` now; partly because it's still in draft mode, so I'm not sure you're already fine with a general review?)

**Comment by ElliotFreebairn** [comment]
> > When you mean uninitialized data, is that for `int colour_count` not being `int colour_count = 0` or were you talking about `m_data`
> 
> the former
> 
> > I also realized that the indexing was incorrect while looking for a fix which will help.
> 
> (I didn't look at the actual code of this PR at all yet, beyond seeing it do a successful `make check` now; partly because it's still in draft mode, so I'm not sure you're already fine with a general review?)

Hi, I think a general review would be good. 

I've got one more file to create unit tests for but maybe it's better to review the code already and learn from my mistakes making it easier to write future unit tests. I'm happy to do whatever

**Comment by stbergmann** [comment]
> Hi, I think a general review would be good.
> 
> I've got one more file to create unit tests for but maybe it's better to review the code already and learn from my mistakes making it easier to write future unit tests. I'm happy to do whatever

OK, I'll see to find time tomorrow morning to look at the code.

File: .gitignore [hunk 6]
```diff
@@ -374,6 +374,6 @@ FodyWeavers.xsd
 
 # Ignore generated binary
 pixelbasher
-
+unittest
```
Comment (stbergmann) [inline .gitignore:377]: I'd keep sections separated by empty lines

File: src/bmp.hpp [hunk 7]
```diff
@@ -92,8 +97,11 @@ class BMP
     BMPInfoHeader m_info_header;
 
     std::vector<std::uint8_t> m_data;
+    std::vector<bool> m_soble_edge_mask;
```
Comment (stbergmann) [inline src/bmp.hpp]: typo `soble` -> `sobel`

File: src/bmp.hpp [hunk 8]
```diff
@@ -43,22 +45,25 @@ struct BMPInfoHeader
...
     void read(const char *filename);
     void write(const char *filename);
     void stamp_name(BMP &stamp);
     static void write_side_by_side(const BMP &diff, const BMP &base, const BMP &target, std::string stamp_location, const char *filename);
     void write_with_filter(const char *filename, std::vector<bool> filter_mask);
+    int calculate_colour_count(Colour to_compare) const;
 
     const std::vector<std::uint8_t> &get_data() const { return m_data; }
     const std::vector<bool> &get_blurred_edge_mask() const { return m_blurred_edge_mask; }
+    const std::vector<bool> &get_filtererd_vertical_edge_mask() const { return m_filtered_vertical_edges; }
```
Comment (stbergmann) [inline src/bmp.hpp]: typo `filtererd` -> `filtered`

File: src/bmp.cpp [hunk 9]
```diff
@@ -37,11 +37,48 @@ BMP::BMP(const char *filename)
     read(filename);
```
Comment (stbergmann) [inline src/bmp.cpp:38]: Before this PR, `read` was a member function that was only used from here (so could have been private, we missed to fix that in the previous PR).  It is OK for a private function to leave its object in an inconsistent state (i.e., the object's invariants are not fulfilled), as long as the invariants are fulfilled when leaving the outermost public function (which had directly or indirectly called the private function as a helper).  This constructor (a public function) calls `read` as a helper and then calculates values for various non-static data members, based on what was read, which presumably establishes the object's invariants.

Now, `read` is also called from test code, but leaving those non-static data members alone, so that likely means that the given object doesn't fulfull its invariants?

To get this disentangled, you need to (a) make up your mind what the invariants are (it can help to write them down as a comment, next to those non-static data members in the class definition that represent those invariants), and (b) make sure that for each public non-static member function, the invariants are fulfilled whenever leaving that function.

File: test/test_bmp.cpp [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+#include "../src/bmp.hpp"
+#include "../src/pixel.hpp"
+
+#include <bit>
+#include <iostream>
+
+using Catch::Matchers::ContainsSubstring;
+
+TEST_CASE("Read BMP files" , "[bmp][read]") {
+    std::string bmp_path;
```
Comment (stbergmann) [inline test/test_bmp.cpp]: If all the `SECTION`s happen to independently each need a var `bmp_path`, I'd not extract that out of those `SECTION`s.  (Otherwise, you might wonder whether in some cases the value assigned to it in one `SECTION` could be relevant still after leaving that `SECITON`.)

File: test/test_bmp.cpp [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+    SECTION("Reads solid_white.bmp successfully") {
+        bmp_path = "test_data/solid_white.bmp";
+        BMP solid_white;
+
+        REQUIRE_NOTHROW(solid_white.read(bmp_path.c_str()));
```
Comment (stbergmann) [inline test/test_bmp.cpp]: * Instead of using `BMP::read`, why not use the corresponding `BMP` constructor?  (Because that can't easily be put into a `REQUIRE_NOTHROW`?  Then see the Catch2 documentation and use a lambda.)
* I'd de-clutter this and directly pass a string literal into the `read` call, getting rid of `bmp_path` and `.c_str()`

**Review by stbergmann — COMMENTED** [review]
Here's a first bunch of review remarks.  (Let me know if we should discuss that "invariants" thing further.)

**Comment by ElliotFreebairn** [comment]
> Here's a first bunch of review remarks. (Let me know if we should discuss that "invariants" thing further.)

Yeah that would be great, I know a small amount about invariants from a couple math courses, but haven't come across it programming yet. 

I'm having a go at trying to deal with the invariance and clean up the bmp structure/the public methods, those changes should be pushed by the end of the afternoon/day

File: .gitignore [hunk 12]
```diff
@@ -374,6 +374,8 @@ FodyWeavers.xsd
 
 # Ignore generated binary
 pixelbasher
+
```
Comment (stbergmann) [inline .gitignore]: Ah, sorry, what I meant was the empty line between the `# Ignore generated binary` and `# Ignore dump folders` sections.  I wouldn't add an additional empty line between the items of this section.

File: test/test_bmp.cpp [hunk 13]
```diff
@@ -11,50 +11,46 @@ TEST_CASE("Read BMP files" , "[bmp][read]") {
     std::string bmp_path;
     SECTION("Reads solid_white.bmp successfully") {
         bmp_path = "test_data/solid_white.bmp";
-        BMP solid_white;
 
-        REQUIRE_NOTHROW(solid_white.read(bmp_path.c_str()));
+        REQUIRE_NOTHROW(new BMP(bmp_path.c_str()));
```
Comment (stbergmann) [inline test/test_bmp.cpp]: That's not a good idea, because now you leak the `BMP` instance on the heap.  But you can just generate a temporary, which will be destroyed at the end of the expression (and don't even need a lambda),
```
REQUIRE_NOTHROW(BMP("test_data/solid_white.bmp")):
```

File: src/bmp.cpp [hunk 14]
```diff
@@ -78,6 +79,14 @@ BMP::BMP(int width, int height, bool has_alpha)
 
     m_info_header.size_image = m_data.size();
     m_file_header.file_size = m_file_header.offset_data + m_data.size();
+
+    // calculate & store invariants
+    m_background_value = get_average_colour();
```
Comment (stbergmann) [inline src/bmp.cpp]: That's a move in the right direction overall.  But take for example `m_background_value`.  It is computed via `get_average_colour`, whose return value depends on the state of `m_data`.  So whenever `m_data` changes (`set_data`!), `m_background_value` would need to be recalculated to keep the invariant "`m_background_value` represents the background value for the bitmap represented by `m_data`" fulfilled.

On the other hand, depending on where and how often `m_background_value` is used, and how expensive it is to compute its value on the fly (via `get_average_colour`), it might be easier to not have a data member `m_background_value` in the first place (so no need to worry that any invariants involving it will be fulfilled, as there simply will be no such invariants), and just compute its value (from the current `m_data`) whenever needed.

File: test/test_bmp.cpp (hunk 13 above)
Comment (ElliotFreebairn) [reply test/test_bmp.cpp]: Oh I didn't know about that, wow, that's very helpful to know, thank you

File: src/bmp.cpp (hunk 14 above)
Comment (ElliotFreebairn) [reply src/bmp.cpp]: Thank you, that make a lot of sense. I've done a full reconstruction of the BMP code, removing the majority of the member variables (which are rarely calculated) and made more methods static and const.

**Comment by ElliotFreebairn** [comment]
Hopefully the invariance and code structure has improved, and thanks for the recent reviews/comments, I'm learning alot :)

**Comment by ElliotFreebairn** [comment]
> Hopefully the invariance and code structure has improved, and thanks for the recent reviews/comments, I'm learning alot :)

I've just realized that a couple of the `get_*_masks`, mainly `get_filtered_vertical_edge_mask`, are getting called multiple times per object. For example the MSO original BMP is compared against a minimum of three BMP's and thus its masks are calculated three times, which is quite costly.

I reckon it will better to store them as member variables and recalculate them whenever `m_data` is updated, or store them in the for loop within `main`

**Comment by ElliotFreebairn** [comment]
Hi @stbergmann 

I've got a quick question about invariance

```
void BMP::write_with_filter(const BMP &base, std::string filename, std::vector<bool> filter_mask)
{
    const int base_width = base.get_width();
    const int base_height = base.get_height();
    BMP copy(base);
    std::vector<uint8_t> copy_data = base.get_data();
    for (int y = 0; y < base_height; y++)
    {
        for (int x = 0; x < base_width; x++)
        {
            int index = y * base_width + x;
            int byte_index = index * 4;

            if (filter_mask[index])
            {
                copy_data[byte_index + 0] = 0;
                copy_data[byte_index + 1] = 0;
                copy_data[byte_index + 2] = 255;
                copy_data[byte_index + 3] = 255;
            }
        }
    }
    copy.set_data(copy_data);
    copy.write(filename);
}
```

In this method, base is passed as an argument (e.g. an MSO image). To apply the filter, I create a copy of base, modify its pixel data, and then write it out to a file. Since this copy is only used inside the method and destroyed at the end of the scope, do I need to maintain all of its invariants (e.g. keeping the masks in sync), or is it alright to leave those unchanged given that the object will not be accessed afterward?


Thanks

File: src/bmp.cpp [hunk 15]
```diff
@@ -30,20 +30,56 @@ const static BMPColourHeader colour_header = {
...
 
-BMP::BMP() {}
+BMP::BMP()
+{
+
+}
 
-void BMP::read(const char *filename)
+// To be used in tests
+BMP::BMP(int width, int height, bool has_alpha)
```
Comment (stbergmann) [inline src/bmp.cpp]: just noticed that `has_alpha` is completely unused and could just be dropped?

File: src/bmp.cpp [hunk 16]
```diff
@@ -30,20 +30,56 @@ const static BMPColourHeader colour_header = {
...
-    m_non_background_count = get_non_background_pixel_count(m_background_value);
-    m_blurred_edge_mask = blur_edge_mask(sobel_edges<245>());
-    m_vertical_edges = filter_long_vertical_edge_runs(get_vertical_edges<245>(), 10);
+
+    m_blurred_edge_mask = blur_edge_mask();
+    m_filtered_edge_mask = filter_long_vertical_edge_runs(10);
 }
 
-BMP::BMP() {}
+BMP::BMP()
```
Comment (stbergmann) [inline src/bmp.cpp]: this constructor is still problematic, as it at last leaves the `m_*_header` data members uninitialized, so that e.g. a following `get_width` would return garbage.  (And the default initialization of the other data members might or might not happen to fulfill the class's invariants---hard to tell for an instance for which even width and height are unspecified :)

The easiest thing might be to allow `width` and `height` for the following constructor to be zero (or would that cause any issues), and then let this constructor forward to that constructor,
```
BMP::BMP(): BMP(0, 0, false) {}
```

File: src/pixelbasher.cpp [hunk 17]
```diff
@@ -66,6 +68,7 @@ BMP PixelBasher::compare_bmps(const BMP &original, const BMP &target, bool enabl
         }
     }
     diff.set_data(diff_data);
+    diff.recalculate_masks();
```
Comment (stbergmann) [inline src/pixelbasher.cpp]: Splitting `set_data` and `recalculate` into two public member functions is problematic, as it leaves the invariants violated after the first call, and only fixes them with the second.  Better to have `set_data` call `recalculate_masks` internally, so that at the return from `set_data` the invariants hold.  (And if that causes unnecessary recalculations on temporary `BMP` instances for the calls to `set_data` internally in `bmp.cpp`, you could add some private `set_data_internal` or `set_data_unsafe` that skips them and can only be called from within `BMP` code itself, so not getting into trouble with upholding the invariants and the boundaries of the `BMP` code.)

File: test/test_bmp.cpp [hunk 18]
```diff
@@ -0,0 +1,271 @@
+#include <catch2/catch_all.hpp>
+#include "../src/bmp.hpp"
+#include "../src/pixel.hpp"
+
+#include <bit>
+#include <iostream>
+
+using Catch::Matchers::ContainsSubstring;
+
```
Comment (stbergmann) [inline test/test_bmp.cpp:5]: I'd also add some very basic tests, like creating a default `BMP` instance, or some select `BMP(width, height)` variant(s), and then checking e.g. `get_width`, `get_heigth` on them.  (That also automatically makes you consider edge cases, like "what is the behavior for a default-initialized `BMP dummy;`?")

**Comment by ElliotFreebairn** [comment]
Hi, thanks for the recent review :) I've pushed the changes which should hopefully resolves the issues.

File: src/bmp.cpp [hunk 19]
```diff
@@ -365,6 +365,17 @@ int BMP::get_background_value() const
 }
 
 void BMP::set_data(std::vector<std::uint8_t> &new_data)
+{
+    if (new_data.size() != m_data.size())
+    {
+        throw std::runtime_error("New data size " + std::to_string(new_data.size()) +
+                                 " differs to current data size " + std::to_string(m_data.size()));
+    }
+    m_data = new_data;
```
Comment (stbergmann) [inline src/bmp.cpp]: I'd not duplicate the code and just call `set_data_internal` here

**Review by stbergmann — APPROVED** [review]
LGTM; did you want to add test code for further classes here? or else you can push it now I'd say

**Comment by ElliotFreebairn** [comment]
> LGTM; did you want to add test code for further classes here? or else you can push it now I'd say

I'll add test code for pixelbasher.cpp and a couple for pixel.hpp. But it should be much easier now given you're reviews and help with the bmp tests

**Comment by ElliotFreebairn** [comment]
Hi, @stbergmann

I was just wondering If there's anything else I could add or do in relation to this tool, after these tests have been merged?

I've only got about about a 7 days left of work, and looking for some stuff to do. Thank you :)

File: test/test_pixel.cpp [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+#include "../src/pixel.hpp"
+
+#include <bit>
+#include <iostream>
+
+using Catch::Matchers::ContainsSubstring;
+
+TEST_CASE("Pixel struct functionality", "[pixel][struct]") {
+    SECTION("Get BGRA values from pixel data") {
+        std::uint8_t pixel_data[] = {255, 0, 0, 255};
```
Comment (stbergmann) [inline test/test_pixel.cpp]: I'd use distinct values for the four components.  Currently, the test would happen to succeed even if the implementation of `get_bgra` accidentally reversed the order in which it processed the four components.

File: test/test_pixel.cpp [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+
+        bool differs = Pixel::differs_from(original_blue, target_green, 128, false);
+        REQUIRE(differs == true);
+    }
+
+     SECTION("Check if two identical pixels differ") {
+        PixelValues original_red = {0, 0, 255, 255};
+        PixelValues target_red = {0, 0, 255, 255};
+
+        bool differs = Pixel::differs_from(original_red, target_red, 128, false);
```
Comment (stbergmann) [inline test/test_pixel.cpp:29]: Maybe have some more "interesting" tests of `differs_from` than just "exactly the same" (with varying values of `background_value`, `near_edge`, `threshold` parameters, even).

File: test/test_bmp.cpp [hunk 22]
```diff
@@ -0,0 +1,295 @@
+#include <catch2/catch_all.hpp>
+#include "../src/bmp.hpp"
+#include "../src/pixel.hpp"
+
+#include <bit>
+#include <iostream>
```
Comment (stbergmann) [inline test/test_bmp.cpp]: you're not actually needing any of these standard header includes (in any of those three `test/test_*.cpp`)?

File: test/test_pixelbasher.cpp [hunk 23]
```diff
@@ -0,0 +1,64 @@
+#include <catch2/catch_all.hpp>
+
+#include "../src/pixelbasher.hpp"
+
+#include <bit>
+#include <iostream>
+
+using Catch::Matchers::ContainsSubstring;
+
+TEST_CASE("PixelBasher compare two BMP's", "[pixelbasher][compare]") {
```
Comment (stbergmann) [inline test/test_pixelbasher.cpp:6]: add a test for `PixelBasher::compare_bmps` with `enable_minor_differences` being `true`?

File: test/test_pixelbasher.cpp [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+    SECTION("Compare two different BMPs and generate a diff image") {
+        BMP original("test_data/solid_white.bmp");
+        BMP target("test_data/vertical_edges.bmp");
+
+        BMP diff = PixelBasher::compare_bmps(original, target, false);
+
+        REQUIRE(diff.get_width() <= original.get_width());
```
Comment (stbergmann) [inline test/test_pixelbasher.cpp]: For all these checks in this file, I wonder if it wouldn't be better if they were testing for more concrete values, e.g.
```
        REQUIRE(diff.get_width() == 50); // <= original.get_width()
```
etc. (leaving that "most general" checking predicate as a comment, as a guidance for the reader and future maintainer, e.g. should the `test_data` files change in the future).  (That way, the tests could find more regressions in the implemenations---if you are sure the concrete actual values you'll put here are correct, that is :)

**Comment by stbergmann** [comment]
> I was just wondering If there's anything else I could add or do in relation to this tool, after these tests have been merged?

Ideas that come to mind are looking at performance,
* <https://www.brendangregg.com/flamegraphs.html>
* <https://kcachegrind.github.io/html/Home.html>

and/or correctness,
* <https://valgrind.org/info/tools.html#memcheck>
* <https://gcc.gnu.org/onlinedocs/gcc/Instrumentation-Options.html#index-fsanitize_003daddress>
* <https://gcc.gnu.org/onlinedocs/gcc/Instrumentation-Options.html#index-fsanitize_003dundefined>
*  <https://lcamtuf.coredump.cx/afl/>

**Review by stbergmann — APPROVED** [review]
LGTM
